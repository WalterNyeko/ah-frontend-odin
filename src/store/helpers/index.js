import { pick, isObjectLike, get, zipObject } from 'lodash';
import { getAllArticles } from 'store/actions/articles';

export const mapStateToProps = props => state => {
  if (Array.isArray(props)) return pick(state, props);

  if (isObjectLike(props)) {
    const keys = Object.keys(props);

    const values = keys.map(key => get(state, props[key]));
    return zipObject(keys, values);
  }

  return {};
};

export const getArticles = ({ articles }) => ({ articles: articles.all });

export const fetAllArticles = dispatch => ({ fetchAllArticles: () => dispatch(getAllArticles()) });
export const sortByUpdatedAt = data => {
  const toDate = object => {
    const updatedAt = object.updated_at || object.updatedAt;
    return (updatedAt && new Date(updatedAt)) || 0;
  };
  // sort articles by updated_at
  return data.sort((a, b) => toDate(b) - toDate(a));
};
export default mapStateToProps;
