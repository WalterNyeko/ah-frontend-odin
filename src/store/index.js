import { createStore, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import Http from 'utils/Http';
import { API_URL } from 'utils/constants';

import rootReducer from './reducers';

const http = new Http(API_URL);
const shouldLog = process.env.NODE_ENV === 'development';
const middleware = [thunk.withExtraArgument(http)];

/* istanbul ignore if */
if (shouldLog) {
  middleware.push(logger);
}

const store = createStore(rootReducer, applyMiddleware(...middleware));

export default store;
