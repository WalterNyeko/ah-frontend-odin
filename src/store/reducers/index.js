import { combineReducers } from 'redux';
import UIReducer from './UIReducer';
import articles from './article';

export default combineReducers({
  UIReducer,
  articles,
});
