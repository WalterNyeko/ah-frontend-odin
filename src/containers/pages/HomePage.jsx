import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { take } from 'lodash';
import DefaultLayout from 'containers/layout/DefaultLayout';
import JoinSection from 'containers/layout/JoinPlatiformSection';
import Articles from 'components/articles/Articles';
import { mapStateToProps, fetAllArticles } from 'store/helpers';

class IndexPage extends React.Component {
  componentWillMount() {
    this.props.fetchAllArticles();
  }

  render() {
    const { articles, user } = this.props;

    return (
      <DefaultLayout className="indexPage">
        <div className="container">
          <h2 className="text-center">Latest Articles</h2>
          <div className="mt-3">
            <Articles articles={take(articles, 4)} user={user} />
          </div>
        </div>
        <JoinSection />
      </DefaultLayout>
    );
  }
}

IndexPage.defaultProps = {
  user: null,
};

IndexPage.propTypes = {
  articles: PropTypes.array.isRequired,
  fetchAllArticles: PropTypes.func.isRequired,
  user: PropTypes.shape(),
};

export default connect(
  mapStateToProps({ articles: 'articles.all', user: 'authentication.user' }),
  fetAllArticles,
)(IndexPage);
