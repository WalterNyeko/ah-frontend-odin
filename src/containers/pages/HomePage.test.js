import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from 'store';
import HomePage from './HomePage';

describe('HomePage', () => {
  test('It renders the same UI', () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <MemoryRouter>
              <HomePage />
            </MemoryRouter>
          </Provider>,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });
});
