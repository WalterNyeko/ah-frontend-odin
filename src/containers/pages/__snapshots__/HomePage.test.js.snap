// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage It renders the same UI 1`] = `
<div
  className="indexPage"
>
  <div
    className="ui-loader"
  >
    <svg
      height="38"
      stroke="#fff"
      viewBox="0 0 38 38"
      width="38"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fillRule="evenodd"
      >
        <g
          strokeWidth="2"
          transform="translate(1 1)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            strokeOpacity=".5"
          />
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
          >
            <animateTransform
              attributeName="transform"
              dur="1s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-primary mb-2"
  >
    <a
      aria-current="page"
      className="navbar-brand d-flex align-items-center active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      <img
        alt="Author Havens Logo"
        src="logo.png"
      />
      <span
        className="h3 font-weight-bold  pl-1"
      >
         Authors Haven 
      </span>
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav"
      >
        <li
          className="nav-item"
        />
      </ul>
      <form
        className="form-inline my-2 ml-auto"
        method="post"
        onSubmit={[Function]}
      />
      <ul
        className="navbar-nav"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/search"
            onClick={[Function]}
          >
            <i
              className="nav-link-icon"
            >
              <svg
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.32 14.9l5.39 5.4a1 1 0 0 1-1.42 1.4l-5.38-5.38a8 8 0 1 1 1.41-1.41zM10 16a6 6 0 1 0 0-12 6 6 0 0 0 0 12z"
                />
              </svg>
            </i>
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            className="nav-link"
            href="/articles"
            onClick={[Function]}
          >
            Articles
          </a>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="nav-link text-light btn-link"
            data-id="login"
            onClick={[Function]}
          >
            login
          </button>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="nav-link text-light btn-link"
            data-id="signup"
            onClick={[Function]}
          >
            Register
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="main"
  >
    <div
      className="container"
    >
      <h2
        className="text-center"
      >
        Latest Articles
      </h2>
      <div
        className="mt-3"
      >
        <div />
      </div>
    </div>
    <section
      className="p-3 bg-primary text-white"
    >
      <div
        className="container-fluid"
      >
        <h3>
          Join Authors Haven
        </h3>
        <p
          className="mt-3"
        >
          A Social platform for like-minded authors to foster inspiration and innovation by leveraging the modern web to enable authors share original content, target articles to an audience that finds them relevant and more.
        </p>
        <div
          className="d-flex justify-content-end"
        >
          <button
            className="btn join-button"
            onClick={[Function]}
            type="submit"
          >
            SignUp
          </button>
        </div>
      </div>
    </section>
  </div>
</div>
`;
